### PR TITLE
Vaapi frame pool

### PR DESCRIPTION
**Pull request by mbouron**

### CODE REVIEW

File: src/decoder_ffmpeg.c [hunk 1]
```diff
@@ -104,6 +104,108 @@ static int init_mediacodec(struct decoder_ctx *ctx)
...
+    {FF_PROFILE_H264_MAIN,                 AV_CODEC_ID_H264, VAProfileH264Main},
+    {FF_PROFILE_H264_HIGH,                 AV_CODEC_ID_H264, VAProfileH264High},
+    {FF_PROFILE_H264_HIGH_10,              AV_CODEC_ID_H264, VAProfileH264High10},
+
+    {FF_PROFILE_HEVC_MAIN,                 AV_CODEC_ID_HEVC, VAProfileHEVCMain},
+    {FF_PROFILE_HEVC_MAIN_10,              AV_CODEC_ID_HEVC, VAProfileHEVCMain10},
+    {FF_PROFILE_HEVC_MAIN_STILL_PICTURE,   AV_CODEC_ID_HEVC, VAProfileHEVCMain},
+};
+
+static int is_entrypoint_supported(VAEntrypoint *entrypoint_list, int num_entrypoints, VAEntrypoint entrypoint)
```
Comment (ubitux) [inline src/decoder_ffmpeg.c]: `entrypoint_list` can be const

File: src/decoder_ffmpeg.c [hunk 2]
```diff
@@ -104,6 +104,108 @@ static int init_mediacodec(struct decoder_ctx *ctx)
...
+    }
+
+    int num_profiles = 0;
+    VAStatus status = vaQueryConfigProfiles(display, profile_list, &num_profiles);
+    if (status != VA_STATUS_SUCCESS) {
+        ret = AVERROR_EXTERNAL;
+        goto end;
+    }
+
+    codecpar = avcodec_parameters_alloc();
```
Comment (ubitux) [inline src/decoder_ffmpeg.c:222]: missing check?

File: src/decoder_ffmpeg.c [hunk 3]
```diff
@@ -104,6 +104,108 @@ static int init_mediacodec(struct decoder_ctx *ctx)
...
+}
+
+static int is_codec_supported(struct decoder_ctx *ctx)
+{
+    int ret = 0;
+    AVCodecContext *avctx = ctx->avctx;
+    AVCodecParameters *codecpar = NULL;
+    VADisplay display = ctx->opaque;
+    int max_num_profiles = vaMaxNumProfiles(display);
+    int max_num_entrypoints = vaMaxNumEntrypoints(display);
```
Comment (ubitux) [inline src/decoder_ffmpeg.c]: could be `const`

**Review by ubitux — COMMENTED** [review]
You may want to add a Changelog entry to mention important VAAPI fixes

File: src/decoder_ffmpeg.c (hunk 1 above)
Comment (mbouron) [reply src/decoder_ffmpeg.c]: Done.

File: src/decoder_ffmpeg.c (hunk 2 above)
Comment (mbouron) [reply src/decoder_ffmpeg.c:222]: Done.

File: src/decoder_ffmpeg.c (hunk 3 above)
Comment (mbouron) [reply src/decoder_ffmpeg.c]: Done.

**Comment by mbouron** [comment]
> You may want to add a Changelog entry to mention important VAAPI fixes

Done in respective commits.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - Fix VAAPI errors that were caused by an undersized surface pool for some video
   streams
+- Fix VAAPI race conditions (leading to crashes) hapenning notably at EOF
```
Comment (ubitux) [inline CHANGELOG.md]: happening

**Review by ubitux — APPROVED** [review]
Should be good with the typo fixed

File: CHANGELOG.md (hunk 4 above)
Comment (mbouron) [reply CHANGELOG.md]: Fixed and reworded a bit the entry.

**Pull request merged**